ICSharpCode.Decompiler/IL/Transforms/InlineExpressionTreeParameterDeclarations.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System.Diagnostics;
+using System.Linq;
+using ICSharpCode.Decompiler.TypeSystem;
+
+namespace ICSharpCode.Decompiler.IL.Transforms
+{
+	class InlineExpressionTreeParameterDeclarationsTransform : IStatementTransform
+	{
+		public void Run(Block block, int pos, StatementTransformContext context)
+		{
+			if (block.Instructions[pos] is StLoc parameterVariableInst) {
+				bool alwaysInlined = true;
+				if (MatchParameterVariableAssignment(parameterVariableInst, out ILVariable v, out ILInstruction init)) {
+					foreach (var LdParameterVariableInst in context.Function.Descendants.OfType<LdLoc>()) {
+						if (LdParameterVariableInst.MatchLdLoc(v)) {
+							ILInstruction parent = LdParameterVariableInst.Parent;
+							while (parent != null) {
+								if (parent is CallInstruction call &&
+									call.Method.FullName.Equals("System.Linq.Expressions.Expression.Lambda") &&
+									!((call.Parent as CallInstruction)?.Method.FullName.Equals("System.Linq.Expressions.LambdaExpression.Compile") ?? false) &&
+									!((call.Parent as CallInstruction)?.Method.FullName.Equals("System.Linq.Expressions.Expression.Compile") ?? false)) {
+									LdParameterVariableInst.ReplaceWith(init.Clone());
+									goto skipAlwaysInlined;
+								}
+								parent = parent.Parent;
+							}
+							alwaysInlined = false;
+							skipAlwaysInlined: ;
+						}
+					}
+					if (alwaysInlined)
+						block.Instructions.RemoveAt(pos);
+				}
+			}
+		}
+
+		bool MatchParameterVariableAssignment(ILInstruction expr, out ILVariable v, out ILInstruction init)
+		{
+			// stloc(v, call(Expression::Parameter, call(Type::GetTypeFromHandle, ldtoken(...)), ldstr(...)))
+			if (!expr.MatchStLoc(out v, out init))
+				return false;
+			if (/*v.HasGeneratedName || */v.Kind == VariableKind.Parameter /*|| v.IsPinned*/) // TODO
```

Comment:
I think `v.Kind != VariableKind.Local && v.Kind != VariableKind.StackSlot` should suffice here.